tensorzero-internal/src/inference/providers/gcp_vertex_gemini.rs

```diff
@@ -426,26 +759,314 @@ impl InferenceProvider for GCPVertexGeminiProvider {
 
     async fn start_batch_inference<'a>(
         &'a self,
-        _requests: &'a [ModelInferenceRequest<'_>],
-        _client: &'a reqwest::Client,
-        _dynamic_api_keys: &'a InferenceCredentials,
+        requests: &'a [ModelInferenceRequest<'_>],
+        http_client: &'a reqwest::Client,
+        dynamic_api_keys: &'a InferenceCredentials,
     ) -> Result<StartBatchProviderInferenceResponse, Error> {
-        Err(ErrorDetails::UnsupportedModelProviderForBatchInference {
-            provider_type: "GCP Vertex Gemini".to_string(),
+        let Some(batch_config) = &self.batch_config else {
+            return Err(ErrorDetails::Config {
+                message: "Missing config section: `[provider_types.gcp_vertex_gemini.batch]`"
+                    .to_string(),
+            }
+            .into());
+        };
+
+        let api_key = self
+            .credentials
+            .get_api_key(&self.audience, dynamic_api_keys)?;
+
+        let mut raw_requests = Vec::with_capacity(requests.len());
+        let mut jsonl_data = Vec::new();
+        for request in requests {
+            let body = GCPVertexGeminiRequest::new(request, &self.model_id)?;
+            let line =
+                serde_json::to_string(&GCPVertexBatchLine { request: body }).map_err(|e| {
+                    Error::new(ErrorDetails::Serialization {
+                        message: format!(
+                            "Error serializing request: {}",
+                            DisplayOrDebugGateway::new(e)
+                        ),
+                    })
+                })?;
+
+            jsonl_data.write_all(line.as_bytes()).map_err(|e| {
+                Error::new(ErrorDetails::Serialization {
+                    message: format!("Error writing to JSONL: {}", DisplayOrDebugGateway::new(e)),
+                })
+            })?;
+            jsonl_data.write_all(b"\n").map_err(|e| {
+                Error::new(ErrorDetails::Serialization {
+                    message: format!("Error writing to JSONL: {}", DisplayOrDebugGateway::new(e)),
+                })
+            })?;
+            raw_requests.push(line);
         }
-        .into())
+
+        let batch_id = Uuid::now_v7();
+        let input_source_url = join_cloud_paths(
+            &batch_config.input_uri_prefix,
+            &format!("tensorzero-batch-input-{batch_id}.jsonl"),
+        );
+
+        // For now, we use the same set of credentials for writing to Google Cloud Storage as we do for invoking
+        // the Vertex API. In the future, we may want to allow configuring a separate set of credentials.
+        let store_and_path =
+            make_gcp_object_store(&input_source_url, &self.credentials, dynamic_api_keys)?;
+
+        store_and_path
+            .store
+            .put(&store_and_path.path, jsonl_data.into())
+            .await
+            .map_err(|e| {
+                // TODO - use better error
```

Comment:
can you actually change this? -- I want to be able to easily localize issues here

Reply:
This is a stale comment - we already wrap the error in a variant/provider message higher up in the call stack.